[#9560] Add a Jmeter test plan to show the performance testing process

amrut-prabhu opened the pull request:
Part of #9560

JMeter GUI should be used for test plan creation and updates. Though it is technically possible to do it in a text editor, it is foolish to do so in most, if not all, cases.

**Notes:**
These can be done in future PRs:
- [x] Refactor to abstract common code into methods and `utils` classes. Try using POST `/databundle` instead of Backdoor.
- [ ] ~Configure jmeter.properties file. Doesn't seem to have any effect through the plugin now. Can use `copyToBin` in `build.gradle`, but that would require an entire file, not just some property values~
- [x] Delete result files routinely? (they don't get overwritten since they are timestamped) - use `gradlew clean` now, but that cleans everything :/
- [x] Remove instances of `org.json.JSONObject` or `org.json.simple.JSONObject`
- [x] Fix Log4J2 missing file error.  (Not sure whether [this](https://libraries.io/github/foragerr/jmeter-gradle-plugin) is still valid)
- [x] Create relevant data in datastore before executing tests
- [x] Remove created data from the datastore
---

**Outline of Solution**

Structure of the test plan:
<img width="165" alt="2019-03-13_16h46_37" src="https://user-images.githubusercontent.com/29021586/54265399-24794a00-45b0-11e9-9053-5a6997786f05.png">

jacoblipech commented:
@amrut-prabhu Is this ready for review?

amrut-prabhu commented:
@jacoblipech Not yet. Ronak is working on a data generation script

amrut-prabhu commented:
@jacoblipech Actually, while he's working on that, can you do a preliminary review of the existing stuff? 
His changes should be independent of the rest, and shouldn't need to modify the current changes anyways (except for docs).

RonakLakhotia commented:
@amrut-prabhu @jacoblipech added the scripts. 👍

amrut-prabhu commented:
@RonakLakhotia can you fix the failed build?

RonakLakhotia commented on src/jmeter/resources/data/studentProfile.json:
```diff
@@ -0,0 +1,103 @@
+{
+  "accounts": {
```
We don't need to commit this file ?

amrut-prabhu commented:
@jacoblipech Ready for Review

@RonakLakhotia Can you take a look as well?

You can try out the scripts by executing the StudentProfile file in the `setup` and `teardown` packages

jacoblipech commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,89 @@
+There are 3 steps involved in doing a performance test:
+
+1. Setting up the data: This is done by executing the relevant Java script in the `setup` package (e.g. `SetupStudentProfileTest`).
+    Executing the scrip will:
```
*script

jacoblipech commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,89 @@
+    1. Create the relevant TEAMMATES JSON data for the datastore. It is stored as `src/jmeter/resources/data/testName.json`.
+    1. Create the relevant CSV Config data for the JMeter test plan. It is stored as `src/jmeter/resources/data/testNameConfig.csv`.
+    1. Create the entities in the datastore using the JSON data. This can be verified by checking your [local datastore](http://localhost:8080/_ah/admin/datastore).
+        > The backend server (`localhost:8080`) must be started for this to happen successfully.
```
must it be built successfully with `npm run build`? or dont need?

amrut-prabhu replied:
Don't need. Just need the backend endpoints to be available

jacoblipech commented on src/jmeter/README.md:
```diff
@@ -0,0 +1,5 @@
+# Continuous Profiling using JMeter
```
is this `readme` file really necessary?

jacoblipech commented on src/jmeter/java/teammates/performance/scripts/create/config/CreateTestConfigData.java:
```diff
@@ -0,0 +1,113 @@
+            throws IOException {
+
+        if (!TestProperties.createTestDataFolder()) {
+            throw new IOException("Test data direcory does not exist");
```
*directory

jacoblipech commented on src/jmeter/java/teammates/performance/scripts/create/data/CreateTestData.java:
```diff
@@ -0,0 +1,63 @@
+    public void writeJsonDataToFile(JSONObject data) throws IOException {
+
+        if (!TestProperties.createTestDataFolder()) {
+            throw new IOException("Test data direcory does not exist");
```
same here

jacoblipech commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,89 @@
+## Performance Test Workflow
+
+Before running a performance test, you need to setup the relevant data in the GAE datastore. 
+To facilitate this, there are Java scripts in the `teammates.performance.scripts` package (`src/jmeter/java/teammates/performance/scripts/` directory).
```
I think either state the directory or the package, no need show both?

jacoblipech commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,89 @@
+
+There are 3 steps involved in doing a performance test:
+
+1. Setting up the data: This is done by executing the relevant Java script in the `setup` package (e.g. `SetupStudentProfileTest`).
```
can you run the script on `Windows`? Cos I tried running the test file on Mac and it throws an error when creating the data directory

wkurniawan07 replied:
More importantly: can you run the script in bash/shell?

amrut-prabhu replied:
> can you run the script on Windows?

It works for me when I run through IntelliJ.
Can you debug and find out why it's not working for you? The data folder is supposed to be created through lnp/util/Test.Properties#createTestDataFolder

amrut-prabhu replied:
> More importantly: can you run the script in bash/shell?

@wkurniawan07 I don't think we can, unless everything is packaged into a jar.

Even if we go to `out/test/classes/teammates/...` and do `javac`, it will not work due to dependencies on external libraries

amrut-prabhu replied:
> More importantly: can you run the script in bash/shell?

@RonakLakhotia Can you look at this? We'll have to create a gradle task

jacoblipech commented on docs/performance-testing.md:
````diff
@@ -0,0 +1,89 @@
+```
+
+If it runs without any errors:
+- A log file is generated as `build/jmeter/jmeter.log`.
````
![image](https://user-images.githubusercontent.com/28522090/54875038-f3abd700-4e31-11e9-9673-1f314f65aeee.png)
It is not normal to see these error messages in the log file right?

amrut-prabhu replied:
@jacoblipech Did you set up the data before running the test?

jacoblipech commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,89 @@
+You can also use the JMeter GUI (`./gradlew jmGui`) to run performance tests. This is particularly useful when debugging or validating that the test does what is expected.
+However, you should not use the GUI to run large scale tests as it is very resource intensive.
+
+> Remember to **disable or remove all `listeners`** if you enabled them while using the GUI since they can have a negative impact on the performance of the test.
```
this works! work on the grammar a bit.
Why does the Gui redirects directly to the studentProfile.xml test?  Shouldn't we be able to specify the test to run using jmGui? Or it is to be build in the future?

amrut-prabhu replied:
> work on the grammar a bit.

Not sure what you're referring to. I'll split the sentence into 2.

> Why does the Gui redirects directly to the studentProfile.xml test? Shouldn't we be able to specify the test to run using jmGui? Or it is to be build in the future?

 Similar to `jmRun`, it's linked to the `testFileDir` and `jmTestFiles` property in `build.gradle`. It opens the first available file.

jacoblipech commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,89 @@
+
+## Generating Test Reports
+
+It is possible to generate HTML summary reports from the JMeter test results.
```
I think can remove this part since the part directly below talked about html files being generated?

amrut-prabhu replied:
I think it's better to always have some kind of intro before subheadings/bullet points 😅

wkurniawan07 commented on src/jmeter/java/teammates/performance/util/TestProperties.java:
```diff
@@ -0,0 +1,79 @@
+
+            TEAMMATES_URL = Url.trimTrailingSlash(prop.getProperty("test.app.url"));
+
+            Properties buildProperties = new Properties();
```
Is this needed?

wkurniawan07 commented on src/jmeter/java/teammates/performance/util/TestProperties.java:
```diff
@@ -0,0 +1,79 @@
+            JMETER_HOME = prop.getProperty("test.jmeter.home").toLowerCase();
+            JMETER_PROPERTIES_PATH = prop.getProperty("test.jmeter.properties", "").toLowerCase();
+
+        } catch (IOException | NumberFormatException e) {
```
Is `NumberFormatException` ever thrown?

wkurniawan07 commented on src/jmeter/java/teammates/performance/util/TestProperties.java:
```diff
@@ -0,0 +1,79 @@
+     * Creates the test data folder if it does not exist.
+     */
+    public static boolean createTestDataFolder() {
+        File testDataDirectory = new File(TestProperties.TEST_DATA_FOLDER);
```
I don't think the qualifier `TestProperties` is needed.

amrut-prabhu commented on src/jmeter/java/teammates/performance/util/BackDoor.java:
```diff
@@ -0,0 +1,259 @@
+import teammates.common.util.JsonUtils;
+
+/**
+ * Used to create API calls to the back-end without going through the UI.
```
It was causing [macker errors](https://travis-ci.org/TEAMMATES/teammates/jobs/508996703#L643). I'll look at it again and see whether there's a better alternative

amrut-prabhu commented on build.gradle:
```diff
@@ -694,6 +702,34 @@ task e2eTests {
+    jMeterExt.metaClass.reportXslt = "$project.projectDir/src/jmeter/resources/jmeter-results-detail-report_21.xsl"
+}
+
+task copyToBin(type: Copy) {
```
It has been fixed in the [snapshot](https://github.com/jmeter-gradle-plugin/jmeter-gradle-plugin/issues/109#issuecomment-414931402) "beta" build. But for now, we won't be using it .

jacoblipech commented:
Thanks @amrut-prabhu It works now. I will give another thorough review tonight! :)

amrut-prabhu commented:
@jacoblipech Can you take another look? I've addressed the review comments, except for the one about running the scripts from command line. This can be done in another PR, since we still need to finalise what process will be used for automating the setting up and running of tests.

jacoblipech commented on src/jmeter/tests/studentProfile.jmx:
```diff
@@ -0,0 +1,229 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jmeterTestPlan version="1.2" properties="5.0" jmeter="5.0 r1840935">
```
I think you should change this to 4.0? since gradlew only supports 4.0? Will be good to mention that part in the  docs too

jacoblipech commented on src/jmeter/resources/log4j2.xml:
```diff
@@ -0,0 +1,112 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
what is the use of this file actually? Cos like most of the info are commented out

amrut-prabhu replied:
This file is required for generating the `jmeter.log` file when using `jmRun` or `jmReport`.

> Cos like most of the info are commented out

I copied this file over from my Jmeter installation. The comments are present in case you want to change the logging settings. Though it's unlikely that we'll ever change the settings, do you think the comments should still be kept?  It would help avoid googling to try to find the relevant property value, and also gives a list of available properties.

jacoblipech commented on src/jmeter/java/teammates/lnp/scripts/create/data/CreateStudentProfileTestData.java:
```diff
@@ -0,0 +1,169 @@
+        privileges.put("courseLevel", courseLevel);
+
+        instructorDetails.put("privileges", privileges);
+        instructors.put("teammates.test.instructor", instructorDetails);
```
Actually why do you need to create these extra info such as instructors, courses, sessions for the student profile test? Is it just to ensure that it works?

RonakLakhotia replied:
Yup that's right

amrut-prabhu commented:
> Why is it necessary to create the dataConfig file as well?

@jacoblipech We need to generate the `testConfig.csv` files to pass parameters to the JMeter test. For example, in `studentProfileTest.jmx` we pass in `email`,`googleId` and `isAdmin` to parameterise HTTP requests for logging in and getting the student profile. 
The `testConfig.csv` file currently has 100 students. So, each thread ("user") in the load test simulates a different user instance, which feels more realistic.

> Anyway, I will work on populating the data after running the gradlew command after this is merged.

Are you referring to the script to enable execution of the Java scripts from the command line?

amrut-prabhu commented:
@wkurniawan07 Would you mind doing another review of this PR?

amrut-prabhu commented:
@wkurniawan07  Yep, we can remove it.
If we need to, we can look at it using the git history. Shouldn't be a problem

wkurniawan07 commented:
@amrut-prabhu see if you can remove the utility classes `Stopwatch` and `StudentEnrollmentGenerator` as well.

amrut-prabhu commented:
@wkurniawan07 I've updated the code and docs to create the JMeter test from Java. One good thing now is that we can parameterise load values (number of users) without having to update it multiple places.

In addition, we do not need the JMeter download (we never actually _needed_ the entire download after switching to the Java API 😕 ). We just need a couple of property files from the `jmeter/bin` folder. I've included the 2 **necessary** properties files in the [latest commit](https://github.com/TEAMMATES/teammates/pull/9561/commits/24415af47ed66ca6bb8cde2e70d83da9a6a16e6f) and updated the corresponding `test.template.properties` values. Let me know if you think this change is not needed.

wkurniawan07 commented on docs/performance-testing.md:
```diff
@@ -0,0 +1,58 @@
+Each new test case must inherit `BaseLNPTestCase`, create a `LNPTestData` instance, and implement all the `abstract` methods from both classes. 
+The L&P test plans are created using the JMeter API in the overridden `generateTestPlan()` method, which is invoked by `BaseLNPTestCase#getLNPTestPlan(boolean)`. 
+This boolean parameter can be set to `true` in the method call to generate the equivalent `.jmx` file, which can be opened in the JMeter GUI.
+By default, the file is saved as `baseLnpTest.jmx` in the root directory of your TEAMMATES project.
```
This entire section is implementation-specific; very easy to get outdated (in fact it has been, now). See if you can reduce the details to be more resilient to changes.
